src/G2/Execution/InstTypes.hs

```diff
@@ -0,0 +1,65 @@
+{-# LANGUAGE  OverloadedStrings, FlexibleContexts #-}
+module G2.Execution.InstTypes (instType) where
+
+
+import G2.Language
+import qualified Data.HashMap.Lazy as HM
+import qualified G2.Language.ExprEnv as E 
+import qualified Data.List as L 
+import Debug.Trace
+import Data.Foldable (foldl')
+import G2.Language.ExprEnv (isSymbolic)
+
+
+
+generateIds :: NameGen -> Type -> (NameGen, Id)
+generateIds ng t = 
+    let 
+        (tn, ng') = freshSeededString "t" ng 
+    in
+    (ng', Id tn t)
+
+newType :: NameGen -> Id -> TypeEnv -> (Type, TypeEnv, NameGen)
+newType ng i te = 
+    let  
+        (tn, ng') = freshSeededString "T" ng 
+        (dc, ng'') = freshSeededString "DC" ng' 
+
+        i_k = typeOf i
+        ty = TyCon tn i_k
+        (ng''', all_ids) = L.mapAccumL generateIds ng'' (argumentTypes $ PresType i_k)
+        
+        tys = ty : map TyVar all_ids
+        tyapps = mkTyApp tys
+        tyfuns = TyFun TyLitInt tyapps
+        tyforall = foldl' (flip TyForAll) tyfuns all_ids
+
+        nadt = DataTyCon 
+                    {bound_ids = all_ids
+                    ,data_cons = [DataCon dc tyforall]}
+        te' = HM.insert tn nadt te 
+    in
+    (ty, te', ng''')
+
+instType :: ASTContainer t Type => NameGen -> State t -> (NameGen, State t)
```

Comment:
Since this is the only exported function, would be good to add a Haddock style comment (a comment that starts with "-- | " before the type signature- it will let the comment show up in documentation.)

Reply:
BTW, did you think we should add some documentation to instType though we already have documentation for instType'?

Reply:
We're replacing the type variables with concrete types, not with symbolic variables, right?